include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
+///       template parameter that has the same signature as `TheNode` that
+///       will be declared later. This allows us to use it as if it was
+///       declared, provided that only the real `TheNode` is used as this
+///       argument.
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename, size_t, bool>
+         class TheNode,
+         typename FinalType,
+         size_t ParentSmallSize,
+         bool ParentHasEntryNode>
+struct NewNodeBaseTCalc {
+  static constexpr bool
+    NoDerivation = std::is_same_v<FinalType, std::false_type>;
+  using NodeType = TheNode<Node,
+                           EdgeLabel,
+                           HasParent,
+                           SmallSize,
+                           FinalType,
+                           ParentSmallSize,
+                           ParentHasEntryNode>;
+  using DerivedType = std::conditional_t<NoDerivation, NodeType, FinalType>;
+  using GenericGraph = GenericGraph<DerivedType,
+                                    ParentSmallSize,
+                                    ParentHasEntryNode>;
+  using ParentType = Parent<GenericGraph, Node>;
+  using Result = std::conditional_t<HasParent, ParentType, Node>;
+};
+} // namespace detail
+
+/// A node type with support for non-trivial edge types.
+template<typename Node,
+         typename EdgeLabel = Empty,
+         bool HasParent = true,
+         size_t SmallSize = 2,
+         typename FinalType = std::false_type,
+         size_t ParentSmallSize = 16,
+         bool ParentHasEntryNode = true>
+class NewNode : public detail::NewNodeBaseTCalc<Node,
+                                                EdgeLabel,
+                                                HasParent,
+                                                SmallSize,
+                                                NewNode,
+                                                FinalType,
+                                                ParentSmallSize,
+                                                ParentHasEntryNode>::Result {
+public:
+  static constexpr bool is_new_node = true;
+  static constexpr bool has_parent = HasParent;
+  using TypeCalc = detail::NewNodeBaseTCalc<Node,
+                                            EdgeLabel,
+                                            HasParent,
+                                            SmallSize,
+                                            NewNode,
+                                            FinalType,
+                                            ParentSmallSize,
+                                            ParentHasEntryNode>;
+  using DerivedType = typename TypeCalc::DerivedType;
+  using Base = typename TypeCalc::Result;
+
+  struct EdgeView {
+    DerivedType &Neighbor;
+    EdgeLabel &Label;
+  };
+  struct OwningEdge {
+    DerivedType *Neighbor;
+    std::unique_ptr<EdgeLabel> Label;
```

Comment:
In the current implementation we have

```c++
/// Data structure for edge labels                                                                                                                                                            
template<typename Node, typename EdgeLabel>
struct Edge : public EdgeLabel {
  Edge(Node *Neighbor) : Neighbor(Neighbor) {}
  Edge(Node *Neighbor, EdgeLabel EL) : EdgeLabel(EL), Neighbor(Neighbor) {}
  Node *Neighbor;
};
```

Why deviate?

Reply:
Because we can't store `Neighbor` inside the edge - we need two distinct edge objects with different (opposite) neighbors and a shared label.

Reply:
I see.